File: ch6/clip5/boj1726/main.py
from collections import deque

def is_range(r, c, m, n):
    return 1 <= r <= m and 1 <= c <= n

m, n = map(int, input().split())
a = [list(map(int, input().split())) for _ in range(m)]
dist = [[[-1] * (n + 1) for _ in range(m + 1)] for _ in range(5)]
EAST, WEST, SOUTH, NORTH = 1, 2, 3, 4

start_r, start_c, start_dir = map(int, input().split())
end_r, end_c, end_dir = map(int, input().split())

dr = [0, 0, 0, 1, -1]
dc = [0, 1, -1, 0, 0]

for i in range(1, m + 1):
    for j in range(1, n + 1):
        for k in range(1, 5):
            dist[k][i][j] = -1

dist[start_dir][start_r][start_c] = 0
q = deque([(start_r, start_c, start_dir)])

while q:
    now_r, now_c, now_dir = q.popleft()

    for k in range(1, 4):
        nr = now_r + dr[now_dir] * k
        nc = now_c + dc[now_dir] * k

        if not is_range(nr, nc, m, n) or a[nr - 1][nc - 1] == 1:
            break

        if dist[now_dir][nr][nc] != -1:
            continue

        dist[now_dir][nr][nc] = dist[now_dir][now_r][now_c] + 1
        q.append((nr, nc, now_dir))

    for k in range(1, 5):
        if k == now_dir:
            continue

        if dist[k][now_r][now_c] != -1:
            continue

        cnt = 0
        if k in (EAST, WEST):
            if now_dir in (NORTH, SOUTH):
                cnt = 1
            else:
                cnt = 2
        elif k in (SOUTH, NORTH):
            if now_dir in (EAST, WEST):
                cnt = 1
            else:
                cnt = 2

        dist[k][now_r][now_c] = dist[now_dir][now_r][now_c] + cnt
        q.append((now_r, now_c, k))

print(dist[end_dir][end_r][end_c])

File: ch6/clip5/boj1726/Main.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int m = sc.nextInt();
        int n = sc.nextInt();
        int[][] a = new int[m + 1][n + 1];
        int[][][] dist = new int[5][m + 1][n + 1];
        final int EAST = 1, WEST = 2, SOUTH = 3, NORTH = 4;
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                a[i][j] = sc.nextInt();
            }
        }
        Point start = new Point(sc.nextInt(), sc.nextInt(), sc.nextInt());
        Point end = new Point(sc.nextInt(), sc.nextInt(), sc.nextInt());

        int[] dr = {0, 0, 0, 1, -1};
        int[] dc = {0, 1, -1, 0, 0};

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                for (int k = 1; k <= 4; k++) {
                    dist[k][i][j] = -1;
                }
            }
        }

        dist[start.dir][start.r][start.c] = 0;
        Queue<Point> q = new LinkedList<>();
        q.add(start);
        while (!q.isEmpty()) {
            Point now = q.poll();
            for (int k = 1; k <= 3; k++) {
                int nr = now.r + dr[now.dir] * k;
                int nc = now.c + dc[now.dir] * k;
                if (!isRange(nr, nc, m, n)) break;
                if (a[nr][nc] == 1) break;
                if (dist[now.dir][nr][nc] != -1) continue;
                dist[now.dir][nr][nc] = dist[now.dir][now.r][now.c] + 1;
                q.add(new Point(nr, nc, now.dir));
            }
            for (int k = 1; k <= 4; k++) {
                if (k == now.dir) continue;
                if (dist[k][now.r][now.c] != -1) continue;
                int cnt = switch (k) {
                    case EAST, WEST -> (now.dir == NORTH || now.dir == SOUTH) ? 1 : 2;
                    case SOUTH, NORTH -> (now.dir == EAST || now.dir == WEST) ? 1 : 2;
                    default -> 0; // error
                };
                dist[k][now.r][now.c] = dist[now.dir][now.r][now.c] + cnt;
                q.add(new Point(now.r, now.c, k));
            }
        }
        System.out.println(dist[end.dir][end.r][end.c]);
    }

    static boolean isRange(int r, int c, int m, int n) {
        return r >= 1 && r <= m && c >= 1 && c <= n;
    }
}

class Point {
    int r, c, dir;

    Point(int r, int c, int dir) {
        this.r = r;
        this.c = c;
        this.dir = dir;
    }
}
